Support column_list in CONTAINS clause

jaspal007 opened the pull request:
### Description
Previously in Babelfish, the feature of fulltext search for simple terms in CONTAINS clause was not supported for multiple columns. It throws syntax error over TDS endpoint. To fix this, a new grammar is added to support the TSQL multiple column Full-text search syntax. Additional validations are added to check the number of columns that can be Full-text indexed along with the validation for the column being queried is Full-text indexed or not. 

### 1. Issue:
- The syntax for multiple columns passed in Fulltext Search TSQL CONTAINS query is not supported in Babelfish. 
On passing multiple columns in the TSQL CONTAINS query, it throws syntax unsupported error. 

  ```
  1>  EXEC sp_babelfish_configure 'babelfishpg_tsql.escape_hatch_fulltext', 'ignore'
  2> GO
  1> CREATE TABLE fts_table (
  2>               id INT NOT NULL,
  3>               col1 TEXT,
  4>               col2 TEXT)
  5> GO
  1> CREATE UNIQUE INDEX uid ON fts_table(id)
  2> GO
  1> CREATE FULLTEXT INDEX ON fts_table(col1, col2) KEY INDEX uid 
  2> GO
  1> SELECT * FROM fts_table WHERE CONTAINS((col1, col2), 'hello')
  2> GO
  Msg 33557097, Level 16, State 1, Server yashneet-cluster-29-jan-16-07-i, Line 1 
  syntax error at or near "("
  ```
- No validation check to ensure that column passed within the TSQL CONTAINS query are fulltext indexed. 
 
     - current output:
         ```
         1> SELECT * FROM fts_table(col3, 'ghi')
         2> GO

         id               col1             col2             col3
         ---------------  ---------------  ---------------  ---------------
                       1  abc               def             ghi
        ```

     - expected output:
        ```
        1> SELECT * FROM fts_table(col3, 'ghi')
        2> GO
        Msg 33557097, Level 16, State 1, Server BABELFISH, Line 1
        Cannot use a CONTAINS or FREETEXT predicate on column 'col3' because it is not full-text indexed.
        ```

### 2. Changes made to fix the issue:
- Grammar support is extended over ANTLR to support multiple columns to be passed in TSQL CONTAINS query.
- The columns are processed over Babelfish to perform a Fulltext search over the simple_term search string for all the columns passed in the query. 
- Validation check is added to ensure all the columns passed in the TSQL CONTAINS query are a part of Fulltext index created over the table. 
- Added a validation check to ensure that a Fulltext index is created over at most 32 columns. 


TASK: BABEL-4239
Authored-by: Jaspal Singh ijaspals@amazon.com
Signed-off-by: Jaspal Singh ijaspals@amazon.com

### 3. Issues Resolved
BABEL-4239

### Test Scenarios Covered ###
* **Use case based -** YES


* **Boundary conditions -** YES


* **Arbitrary inputs -** YES


* **Negative test cases -** YES


* **Major version upgrade tests -** YES


* **Performance tests -** YES



### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
Support column_list in CONTAINS clause

Previously in Babelfish, the feature of fulltext search for simple terms in CONTAINS clause was not supported for multiple columns. The TSQL syntax of Fulltext search for simple terms in CONTAINS clause over multiple columns was unidentifiable over TDS endpoint of Babelfish and would throw syntax error. To fix this, we have created a new grammar to support the TSQL multiple column Fulltext search syntax for simple terms in CONTAINS clause over TDS endpoint of Bablefish. The support for multiple column fulltext search over CONTAINS clause is extended with validation check over columns if they are a part of existing fulltext index. The fulltext index can be created over 32 columns for a single table.

TASK: BABEL-4239
Signed-off-by: Jaspal Singh ijaspals@amazon.com

Commit:
Support column_list in CONTAINS clause

Previously in Babelfish, the feature of fulltext search for simple terms in CONTAINS clause was not supported for multiple columns. The TSQL syntax of Fulltext search for simple terms in CONTAINS clause over multiple columns was unidentifiable over TDS endpoint of Babelfish and would throw syntax error. To fix this, we have created a new grammar to support the TSQL multiple column Fulltext search syntax for simple terms in CONTAINS clause over TDS endpoint of Bablefish. The support for multiple column fulltext search over CONTAINS clause is extended with validation check over columns if they are a part of existing fulltext index. The fulltext index can be created over 32 columns for a single table.

TASK: BABEL-4239
Signed-off-by: Jaspal Singh ijaspals@amazon.com

Commit:
Support column_list in CONTAINS clause

Previously in Babelfish, the feature of fulltext search for simple terms in CONTAINS clause was not supported for multiple columns. The TSQL syntax of Fulltext search for simple terms in CONTAINS clause over multiple columns was unidentifiable over TDS endpoint of Babelfish and would throw syntax error. To fix this, we have created a new grammar to support the TSQL multiple column Fulltext search syntax for simple terms in CONTAINS clause over TDS endpoint of Bablefish. The support for multiple column fulltext search over CONTAINS clause is extended with validation check over columns if they are a part of existing fulltext index. The fulltext index can be created over 32 columns for a single table.

TASK: BABEL-4239
Signed-off-by: Jaspal Singh ijaspals@amazon.com

jaspal007 committed:
Merge branch 'babelfish-for-postgresql:BABEL_5_X_DEV' into BABEL_4239-5-X

jaspal007 committed:
Merge branch 'babelfish-for-postgresql:BABEL_5_X_DEV' into BABEL_4239-5-X

jaspal007 committed:
Merge branch 'babelfish-for-postgresql:BABEL_5_X_DEV' into BABEL_4239-5-X

roshan0708 commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -790,25 +790,55 @@ is_json_query(List *name)
-    to_tsquery_call = makeToTSQueryFuncCall(search_expr, result_pgconfig);
-    
-    fts = makeA_Expr(AEXPR_OP, list_make1(makeString("@@")), to_tsvector_call, to_tsquery_call, -1);
+static A_Expr *
```
nit: please add some small comment for these helper functions

jaspal007 replied:
Added the comments regarding the use of helper functions

roshan0708 commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1983,7 +1985,173 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+ 			result = DirectFunctionCall1(pg_get_indexdef, ObjectIdGetDatum(indexOid));
+ 			if(DatumGetPointer(result) == NULL)
+ 				return NULL;
+ 			idx = text_to_cstring(DatumGetTextP(result));
```
Please free this pointer.

roshan0708 replied:
Please free pointer `idx` also wherever required.

roshan0708 commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -2184,6 +2352,20 @@ char
 
 	initStringInfo(&query);
 
+	if(column_name->length>32)
```
Why are there 2 checks for column list?

roshan0708 commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -7416,27 +7417,50 @@ void process_execsql_destination(TSqlParser::Dml_statementContext *ctx, PLtsql_s
-                if (check_freetext_predicate(pred->search_condition()))
-                    return true;
-            }
+		if (pred && pred->predicate() && pred->predicate()->freetext_predicate())
```
nit: indentation

roshan0708 commented on test/JDBC/input/full_text_search/fts-contains-vu-prepare.mix:
```diff
@@ -3091,6 +3091,287 @@ go
+GO
+
+-- Create schema
+CREATE SCHEMA fts_schema
```
Let's re-use the older schema for testing

roshan0708 commented on test/JDBC/input/full_text_search/fts-contains-vu-prepare.mix:
```diff
@@ -3091,6 +3091,287 @@ go
 go
 
+-- Create the test table
+CREATE TABLE test_tb (
```
Nit: I'd suggest to use more test oriented names for test tables

roshan0708 commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1983,7 +1985,173 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+ * These columns can be of any text based datatype, like: TEXT, CHAR, VARCHAR, VARBINARY, etc.
+ */
+List
+*get_columns(char *index_stmt)
```
We should validate that `index_stmt` is not `NULL` before further processing.

roshan0708 commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1983,7 +1985,173 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+*get_columns(char *index_stmt)
+{
+ 	List	 *column_name = NULL;
+ 	// Find the starting point after "to_tsvector"
```
nit: try to use `/* */` comment style rather than `//`

roshan0708 commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1983,7 +1985,173 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+ 		StringInfoData bufStr;
+ 
+ 		initStringInfo(&bufStr);
+ 		indexdf += strlen("regconfig, ");
```
We can calculate length of `"regconfig, "` and `"replace_special_chars_fts("` once and reuse them.

roshan0708 commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1983,7 +1985,173 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+			i++;
+		}
+ 		// Extract until the closing parenthesis
+ 		while (indexdf[i] != ')' ) 
```
We can use `strcspn` to find the closing `)`

jaspal007 replied:
used `strchr` to fetch the `column_name`

roshan0708 commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1983,7 +1985,173 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+char
+*toLower(char *str)
+{
+ 	for(int k = 0; k < strlen(str); k++)
```
please add null check here

roshan0708 commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1983,7 +1985,173 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+		return false;
+
+	ft_column_name = get_fulltext_indexed_columns(relid, ft_index_name);
+	if(ft_column_name == NULL)
```
Let's free `ft_index_name` pointer as well since we don't use it anymore after fetching indexed columns info.

roshan0708 commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1983,7 +1985,173 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+ 	}
+	
+	RelationClose(relation);
+	return;
```
This `return` is redundant, we can remove this.

roshan0708 commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1983,7 +1985,173 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+ 		if (strcmp(name, ft_index_name) == 0)
+ 		{
+ 			result = DirectFunctionCall1(pg_get_indexdef, ObjectIdGetDatum(indexOid));
+ 			if(DatumGetPointer(result) == NULL)
```
Please handle closing relation and freeing pointers here as well.

roshan0708 commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1983,7 +1985,173 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+		return false;
+				
+
+	check_column_list(relid, column_name);
```
We should perform this check before fetching fulltext index details. Also, making `check_column_list` a boolean function and handling errors in the main function would improve code flow and readability.

jaspal007 replied:
the check for `column_name_list` is just a precautionary check, to counter any index definition change in future.

roshan0708 commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -790,25 +790,55 @@ is_json_query(List *name)
+}
 
-    return (Node *)fts;
+static Node *
```
`createClause` and `concatClause` are too generic. Should be `createTSMatchExpr` and `createTSOrExpr`.

roshan0708 commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -790,25 +790,55 @@ is_json_query(List *name)
+		Node * query = makeToTSQueryFuncCall(search_expr, result_pgconfig);
+		Node * vec = makeToTSVectorFuncCall((colId->elements[i]).ptr_value, yyscanner, result_pgconfig);
+		column_clause = createClause(vec, query);
+		if(fts != NULL)
+		{
+			fts = concatClause((Node *) fts, (Node *) column_clause);
+		}
+		else
+		{
+			fts = (Node *)column_clause;
+		}
+	}
```
Let's simplify this logic using a ternary operator.

roshan0708 commented:
Fix: use descriptive variable names and consistent spacing/indentation.

Commit:
Support column_list in CONTAINS clause

Previously in Babelfish, the feature of fulltext search for simple terms in CONTAINS clause was not supported for multiple columns. The TSQL syntax of Fulltext search for simple terms in CONTAINS clause over multiple columns was unidentifiable over TDS endpoint of Babelfish and would throw syntax error. To fix this, we have created a new grammar to support the TSQL multiple column Fulltext search syntax for simple terms in CONTAINS clause over TDS endpoint of Bablefish. The support for multiple column fulltext search over CONTAINS clause is extended with validation check over columns if they are a part of existing fulltext index. The fulltext index can be created over 32 columns for a single table.

TASK: BABEL-4239
Signed-off-by: Jaspal Singh ijaspals@amazon.com

jaspal007 committed:
Merge branch 'babelfish-for-postgresql:BABEL_5_X_DEV' into BABEL_4239-5-X

thephantomthief commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -790,25 +790,56 @@ is_json_query(List *name)
+* to_tsvector(pgconfig, col_1) @@ to_tsquery(pgconfig, babelfish_fts_rewrite('<contains_search_condition>') OR
+* to_tsvector(pgconfig, col_2) @@ to_tsquery(pgconfig, babelfish_fts_rewrite('<contains_search_condition>') OR
+* ...
+* * to_tsvector(pgconfig, col_n) @@ to_tsquery(pgconfig, babelfish_fts_rewrite('<contains_search_condition>')
```
Nit: Extra `*` at the start

thephantomthief commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -790,25 +790,56 @@ is_json_query(List *name)
+	A_Expr *column_clause;
+	Node *result_pgconfig;
+	List *args_pgconfig;
+	int colSize = colId->length;
```
Rather than colSize, this seems to be number of cols. Also use `list_length()` instead of `->length`

thephantomthief commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -790,25 +790,56 @@ is_json_query(List *name)
+ 
+	args_pgconfig = list_make1(search_expr);
+	result_pgconfig = (Node *) makeFuncCall(TsqlSystemFuncName("babelfish_fts_contains_pgconfig"), args_pgconfig, COERCE_EXPLICIT_CALL, -1);
+	for(int i = 0; i < colSize; i++)
+	{
+		Node * query = makeToTSQueryFuncCall(search_expr, result_pgconfig);
+		Node * vec = makeToTSVectorFuncCall((colId->elements[i]).ptr_value, yyscanner, result_pgconfig);
+		column_clause = createTSMatchExpr(vec, query);
+		
+		fts = (fts != NULL)? createTSOrExpr((Node *) fts, (Node *) column_clause) : (Node *)column_clause;
+	}
```
Use foreach loop here

thephantomthief commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1956,16 +1956,18 @@ gen_schema_name_for_fulltext_index(const char *schema_name)
+	Oid		schemaOid;
+	Oid		relid;
+	List 		*ft_column_name_list = NIL;
+	int 		len = column_name->length;
```
Use list_length()

thephantomthief commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1975,15 +1977,204 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+
+	ft_column_name_list = get_fulltext_indexed_columns(relid, ft_index_name);
+	/* A safety check to handle the case when the index definition changes from the current definition */
+	if(ft_column_name_list == NULL)
```
change this to `ft_column_name_list == NIL`

thephantomthief commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1975,15 +1977,204 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+	}
+
+	/* Check if column exists */
+	for(int i = 0; i < column_name->length; i++)
```
Use foreach loop

thephantomthief commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1975,15 +1977,204 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+	/* Check if column exists */
+	for(int i = 0; i < column_name->length; i++)
+ 	{
+		if(!check_column_list(relid, (char *)(column_name->elements[i]).ptr_value))
```
Store `(char *)(column_name->elements[i]).ptr_value` in a variable and re use it in check and error message

thephantomthief commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1975,15 +1977,204 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+	}
+
+	/* Check if column is fulltext indexed */
+	for(int i = 0; i < len; i++)
```
Use foreach loop

thephantomthief commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1975,15 +1977,204 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+		bool flag = false;
+		for(int j = 0; j < ft_column_name_list->length; j++)
+		{
+			if(strcmp(toLower((char *)(column_name->elements[i]).ptr_value), toLower((char *)(ft_column_name_list->elements[j]).ptr_value)) == 0)
```
Use `pg_strcasecmp()` instead

thephantomthief commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1975,15 +1977,204 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+}
+
+ 
+/* Check if the columns provided in the freetext predicate query exist or not
```
Comment style should be:

```
/*
 * Start your comment here
 * ...
 */
```

thephantomthief commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1975,15 +1977,204 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+{
+ 	Relation 	relation = RelationIdGetRelation(relid);
+ 	TupleDesc	tupledesc = RelationGetDescr(relation);
+	for (int j = 0; j < tupledesc->natts; j++) 
```
Don't do implicit declaration. Declare `j` at beginning of function.

thephantomthief commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1975,15 +1977,204 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+		if (!attr->attisdropped) 
+		{ 
+			char *str = NameStr(attr->attname);
+			if(strcmp(toLower(column_name), toLower(str)) == 0)
```
use `pg_strcasecmp()` instead

thephantomthief commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1975,15 +1977,204 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+		return NULL;
+	}
+
+ 	while ((indexdf = strstr(indexdf, "regconfig, ")) != NULL) 
```
can case be different?

jaspal007 replied:
as per the previous and current definition, the case is can not differ

thephantomthief commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1975,15 +1977,204 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+		column_name = pstrdup(indexdf);
+
+		appendStringInfoString(&bufStr, (const char *)column_name);
+ 		appendStringInfoChar(&bufStr, '\0');
```
Not needed

thephantomthief commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1975,15 +1977,204 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+ * This ensures that case insesitivity.
+ */
+char
+*toLower(char *str)
```
We already have `str_tolower()`

thephantomthief commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -2184,6 +2375,13 @@ char
 
 	initStringInfo(&query);
 
+	if(column_name->length>32)
```
Use `list_length()`

Commit:
Support column_list in CONTAINS clause

Previously in Babelfish, the feature of fulltext search for simple terms in CONTAINS clause was not supported for multiple columns. The TSQL syntax of Fulltext search for simple terms in CONTAINS clause over multiple columns was unidentifiable over TDS endpoint of Babelfish and would throw syntax error. To fix this, we have created a new grammar to support the TSQL multiple column Fulltext search syntax for simple terms in CONTAINS clause over TDS endpoint of Bablefish. The support for multiple column fulltext search over CONTAINS clause is extended with validation check over columns if they are a part of existing fulltext index. The fulltext index can be created over 32 columns for a single table.

TASK: BABEL-4239
Signed-off-by: Jaspal Singh ijaspals@amazon.com

Commit:
Support column_list in CONTAINS clause

Previously in Babelfish, the feature of fulltext search for simple terms in CONTAINS clause was not supported for multiple columns. The TSQL syntax of Fulltext search for simple terms in CONTAINS clause over multiple columns was unidentifiable over TDS endpoint of Babelfish and would throw syntax error. To fix this, we have created a new grammar to support the TSQL multiple column Fulltext search syntax for simple terms in CONTAINS clause over TDS endpoint of Bablefish. The support for multiple column fulltext search over CONTAINS clause is extended with validation check over columns if they are a part of existing fulltext index. The fulltext index can be created over 32 columns for a single table.

TASK: BABEL-4239
Signed-off-by: Jaspal Singh ijaspals@amazon.com

roshan0708 commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -790,25 +790,56 @@ is_json_query(List *name)
 * ... CONTAINS(column_name, '<contains_search_condition>') ...
```
nit: Please update this comment to specify the supported syntax. Something like this - `CONTAINS( { column_name | ( column_list ) }, '<contains_search_condition>')`

roshan0708 commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -790,25 +790,56 @@ is_json_query(List *name)
+{
+	A_Expr *fts;
+	fts = makeA_Expr(AEXPR_OP, list_make1(makeString("@@")), lexpr, rexpr, -1);
+	return fts;
```
nit: This could be simplified with just a return statement.

roshan0708 commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -790,25 +790,56 @@ is_json_query(List *name)
+{
+	Node *fts;
+	fts = makeOrExpr(lexpr, rexpr, -1);
+	return fts;
```
nit: Could be simplified using just a return statement.

roshan0708 commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1975,15 +1977,194 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+	/* Find the starting point after "to_tsvector" */
+	const char* indexdf = index_stmt;
+	int regconfig = strlen("regconfig, ");
+	int replace_special_chars_fts = strlen("replace_special_chars_fts(");
```
In general, `size_t` should be used whenever we're checking size of something.

roshan0708 replied:
Also, please use better variable names here for storing lengths.

roshan0708 commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1975,15 +1977,194 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+*get_columns(char *index_stmt)
+{
+	
+	List	 *column_name_list = NULL;
```
nit: indentation

roshan0708 commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1975,15 +1977,194 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+		 */
+		if((strstr(indexdf, "replace_special_chars_fts(")) != NULL)
+		{
+			indexdf = strstr(indexdf, "replace_special_chars_fts(");
```
Rather than fetching the pointer again, we can use `(start_ptr = strstr(indexdf, "replace_special_chars_fts(")) != NULL` and then reuse this pointer.

roshan0708 commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -7437,27 +7438,50 @@ void process_execsql_destination(TSqlParser::Dml_statementContext *ctx, PLtsql_s
+					{
+						if(col && col->column_name)
+						{
+							string str = getFullText(col->column_name);
```
let's change this to `std::string.... = ::getFullText(...)` for code uniformity.

roshan0708 commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -7437,27 +7438,50 @@ void process_execsql_destination(TSqlParser::Dml_statementContext *ctx, PLtsql_s
+										errmsg("Incorrect syntax near \'%s\'.",
+											str.c_str())));
+							}
+							char * c_string  = pstrdup(str.c_str());
```
nit: extra space after `*` and also use proper variable names.

roshan0708 commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1975,15 +1977,194 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+	if(ft_index_name == NULL)
+		return false;
+
+	ft_column_name_list = get_fulltext_indexed_columns(relid, ft_index_name);
+	/* A safety check to handle the case when the index definition changes from the current definition */
+	if(ft_column_name_list == NIL)
+	{
+		pfree(ft_index_name);
+		return false;
+	}
```
Fix: Validate column existence before full-text index check

roshan0708 commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1975,15 +1977,194 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+ 
+/* 
+ * Check if the columns provided in the freetext predicate query exist or not
+ * The check for the columns existence is handled in BISON, but we require to 
```
Fix: Update the comment. Bison only parses the query, the actual check for column existence is done during execution time from PG side.

roshan0708 commented on test/JDBC/input/full_text_search/fts-contains-vu-verify.mix:
```diff
@@ -546,6 +543,156 @@ go
+GO
+
+-- negative test
+SELECT * FROM fts_contains_multicol_t WHERE contains((txt1, txt2), '')
```
This is not a negative test but `NULL` test

roshan0708 commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1975,15 +1977,194 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+		{
+			ereport(ERROR,
+				(errcode(ERRCODE_INVALID_OBJECT_DEFINITION),
+					errmsg("column \"%s\" does not exist.",
```
Nit: please remove `.` from the error message.

roshan0708 commented on test/JDBC/input/full_text_search/fts-contains-vu-verify.mix:
```diff
@@ -546,6 +543,156 @@ go
+GO
+
+-- negative test
+SELECT * FROM new_schema_fts.multicol_tb WHERE contains((
```
Also not a negative test but NULL test

roshan0708 commented on test/JDBC/input/full_text_search/fts-contains-vu-verify.mix:
```diff
@@ -546,6 +543,156 @@ go
 go
 
+
+-- Multiple column Test cases for the relation fts_contains_multicol_t
```
Let's also add test case for `32` full-text index limit.

roshan0708 commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -2184,6 +2372,13 @@ char
+ 	{
+ 		ereport(ERROR,
+ 			(errcode(ERRCODE_INVALID_OBJECT_DEFINITION),
+ 				errmsg("Fulltext Index on more than 32 columns is currently not supported in Babelfish")));
```
Let's align this message as we discussed: `"Full-text index creation with more than 32 columns is not currently supported in Babelfish."`

Commit:
Support column_list in CONTAINS clause

Previously in Babelfish, the feature of fulltext search for simple terms in CONTAINS clause was not supported for multiple columns. The TSQL syntax of Fulltext search for simple terms in CONTAINS clause over multiple columns was unidentifiable over TDS endpoint of Babelfish and would throw syntax error. To fix this, we have created a new grammar to support the TSQL multiple column Fulltext search syntax for simple terms in CONTAINS clause over TDS endpoint of Bablefish. The support for multiple column fulltext search over CONTAINS clause is extended with validation check over columns if they are a part of existing fulltext index. The fulltext index can be created over 32 columns for a single table.

TASK: BABEL-4239
Signed-off-by: Jaspal Singh ijaspals@amazon.com

jaspal007 committed:
Merge branch 'babelfish-for-postgresql:BABEL_5_X_DEV' into BABEL_4239-5-X

roshan0708 commented:
Please keep the empty newline at the end of test files.

Commit:
Support column_list in CONTAINS clause

Previously in Babelfish, the feature of fulltext search for simple terms in CONTAINS clause was not supported for multiple columns. The TSQL syntax of Fulltext search for simple terms in CONTAINS clause over multiple columns was unidentifiable over TDS endpoint of Babelfish and would throw syntax error. To fix this, we have created a new grammar to support the TSQL multiple column Fulltext search syntax for simple terms in CONTAINS clause over TDS endpoint of Bablefish. The support for multiple column fulltext search over CONTAINS clause is extended with validation check over columns if they are a part of existing fulltext index. The fulltext index can be created over 32 columns for a single table.

TASK: BABEL-4239
Signed-off-by: Jaspal Singh ijaspals@amazon.com

Commit:
Merge remote-tracking branch 'refs/remotes/origin/BABEL_4239-5-X' into BABEL_4239-5-X

thephantomthief commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1975,15 +1977,206 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
 	
+	/* Check if column exists */
+	foreach(column_name, column_name_list)
+		{
```
nit: incorrect indentation

thephantomthief commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1975,15 +1977,206 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+			char *ft_col_name = (char *)(ft_column_name)->ptr_value;
+			if(pg_strcasecmp(col_name, ft_col_name) == 0)
+			{
+				flag = true;
```
why do we need flag here? We can directly throw error here

jaspal007 replied:
Need to check the column name with all the columns present in full-text index to verify if its full text indexed or not. If not, we throw column not full text indexed error.

Commit:
Support column_list in CONTAINS clause

Previously in Babelfish, the feature of fulltext search for simple terms in CONTAINS clause was not supported for multiple columns. The TSQL syntax of Fulltext search for simple terms in CONTAINS clause over multiple columns was unidentifiable over TDS endpoint of Babelfish and would throw syntax error. To fix this, we have created a new grammar to support the TSQL multiple column Fulltext search syntax for simple terms in CONTAINS clause over TDS endpoint of Bablefish. The support for multiple column fulltext search over CONTAINS clause is extended with validation check over columns if they are a part of existing fulltext index. The fulltext index can be created over 32 columns for a single table.

TASK: BABEL-4239
Signed-off-by: Jaspal Singh ijaspals@amazon.com

thephantomthief reviewed (commented):
Please add the upgrade test name in latest schedule file

jaspal007 commented:
> Please add the upgrade test name in latest schedule file

the test cases are added in the existing file itself.
